### PR TITLE
More least squares solvers

### PR DESCRIPTION
**Pull request by tachella**
- I've added new least squares solvers ([LSQR ](https://docs.scipy.org/doc/scipy/reference/generated/scipy.sparse.linalg.lsqr.html)& [BiCGStab](https://en.wikipedia.org/wiki/Biconjugate_gradient_stabilized_method)) that can often outperform the conjugate gradient solver that is already implemented.
 The solvers are mostly used for computing `physics.A_dagger `and `physics.prox_l2` of `LinearPhysics`.
LSQR seems to outperform significantly conjugate gradient in some inverse problems.

- I've fixed the sinc windowing formula and added the option of a windowed anti-aliasing sinc for `physics.Downsampling`. 

### Checks to be done before submitting your PR
- [x] `python3 -m pytest tests/` runs successfully.
- [x] `black .` runs successfully.
- [x] `make html` runs successfully (in the `docs/` directory).
- [x] Updated docstrings related to the changes (as applicable).
- [x] Added an entry to the [CHANGELOG.rst](../CHANGELOG.rst).

### CODE REVIEW

File: deepinv/optim/utils.py [hunk 1]
```diff
@@ -49,47 +190,378 @@ def conjugate_gradient(
...
+        rsnew = dot(r, r, dim=dim).real
+        if torch.all(rsnew < tol):
             if verbose:
-                print(f"Conjugate gradient converged after {_} iterations")
+                print("CG Converged at iteration", _)
+            flag = False
             break
         p = r + p * (rsnew / (rsold + eps))
         rsold = rsnew
+        # print(rsnew[0].item())
```
Comment (matthieutrs) [inline deepinv/optim/utils.py:231]:
```suggestion
```

File: deepinv/optim/utils.py [hunk 2]
```diff
@@ -49,47 +190,378 @@ def conjugate_gradient(
...
+    else:
+        dim = [i for i in range(b.ndim) if i not in parallel_dim]
+
+    if init is not None:
+        x = init
+    else:
+        x = zeros_like(b)
+
+    r = b - A(x)
+    r_hat = r.clone()  # torch.ones_like(r) #torch.rand_like(r) #r.clone()
```
Comment (matthieutrs) [inline deepinv/optim/utils.py:285]:
```suggestion
    r_hat = r.clone()
```

File: deepinv/optim/utils.py [hunk 3]
```diff
@@ -49,47 +190,378 @@ def conjugate_gradient(
...
+        print("BiCGSTAB did not converge")
 
     return x
 
 
+def _sym_ortho(a, b):
+    """
+    Stable implementation of Givens rotation.
+
+    adapted from https://github.com/scipy/scipy/blob/v1.15.1/scipy/sparse/linalg/_isolve/lsqr.py
```
Comment (matthieutrs) [inline deepinv/optim/utils.py:327]:
```suggestion
    Adapted from https://github.com/scipy/scipy/blob/v1.15.1/scipy/sparse/linalg/_isolve/lsqr.py
```

File: deepinv/optim/utils.py [hunk 4]
```diff
@@ -49,47 +190,378 @@ def conjugate_gradient(
...
+                    [vi * alpha.view(bi_shape) for vi, bi_shape in zip(v, b_shape)]
+                )
+            else:
+                return v * alpha.view(b_shape)
+        else:
+            return v * alpha.view(Atb_shape)
+
+    # m = b.size(0)
+    # n = xt.numel()/xt.size(0)
+    # var = torch.zeros_like(xt) if calc_var else None
```
Comment (matthieutrs) [inline deepinv/optim/utils.py:438]:
```suggestion
```

File: deepinv/optim/utils.py [hunk 5]
```diff
@@ -49,47 +190,378 @@ def conjugate_gradient(
...
+        else:
+            eta_sqrt = torch.tensor(eta, device=device).sqrt()
+
+    # ctol = 1 / conlim if conlim > 0 else 0
+    anorm = 0.0
+    acond = torch.zeros(1, device=device)
+    dampsq = eta
+    ddnorm = 0.0
+    # res2 = 0.0
+    # xnorm = 0.0
```
Comment (matthieutrs) [inline deepinv/optim/utils.py:452]:
```suggestion
```

File: deepinv/optim/utils.py [hunk 6]
```diff
@@ -49,47 +190,378 @@ def conjugate_gradient(
...
+        else:
+            rhobar1 = rhobar
+            psi = 0.0
+
+        cs, sn, rho = _sym_ortho(rhobar1, beta)
+        theta = sn * alpha
+        rhobar = -cs * alpha
+        phi = cs * phibar
+        phibar = sn * phibar
+        # tau = sn * phi
```
Comment (matthieutrs) [inline deepinv/optim/utils.py:520]:
```suggestion
```

File: deepinv/optim/utils.py [hunk 7]
```diff
@@ -49,47 +190,378 @@ def conjugate_gradient(
...
+        t1 = phi / rho
+        t2 = -theta / rho
+        dk = scalar(w, 1 / rho, b_domain=False)
+
+        x = x + scalar(w, t1, b_domain=False)
+        w = v + scalar(w, t2, b_domain=False)
+        ddnorm = ddnorm + normf(dk) ** 2
+
+        # if calc_var:
+        #    var = var + dk ** 2
```
Comment (matthieutrs) [inline deepinv/optim/utils.py:531]:
```suggestion
```

File: deepinv/optim/utils.py [hunk 8]
```diff
@@ -49,47 +190,378 @@ def conjugate_gradient(
...
+
+        delta = sn2 * rho
+        gambar = -cs2 * rho
+        rhs = phi - delta * z
+        # zbar = rhs / gambar
+        # xnorm = torch.sqrt(xxnorm + zbar ** 2)
```
Comment (matthieutrs) [inline deepinv/optim/utils.py:537]:
```suggestion
```

File: deepinv/optim/utils.py [hunk 9]
```diff
@@ -49,47 +190,378 @@ def conjugate_gradient(
...
+        gamma = torch.sqrt(gambar**2 + theta**2)
+        cs2 = gambar / gamma
+        sn2 = theta / gamma
+        z = rhs / gamma
+        xxnorm = xxnorm + z**2
+
+        acond = anorm * torch.sqrt(ddnorm).mean()
+        rnorm = torch.sqrt(phibar**2 + psi**2)
+        # arnorm = alpha * abs(tau)
```
Comment (matthieutrs) [inline deepinv/optim/utils.py:546]:
```suggestion
```

File: deepinv/optim/utils.py [hunk 10]
```diff
@@ -49,47 +190,378 @@ def conjugate_gradient(
...
+                b_shape[j].append(b[j].shape[i] if i in parallel_dim else 1)
+    else:
+        for i in range(len(b.shape)):
+            b_shape.append(b.shape[i] if i in parallel_dim else 1)
+
+    Atb_shape = []
+    for i in range(len(xt.shape)):
+        Atb_shape.append(xt.shape[i] if i in parallel_dim else 1)
+
+    def scalar(v, alpha, b_domain):
```
Comment (matthieutrs) [inline deepinv/optim/utils.py:425]: Should this be a standalone function in its own right, similar to `dot`?

File: deepinv/optim/utils.py [hunk 11]
```diff
@@ -49,47 +190,378 @@ def conjugate_gradient(
...
+    if torch.all(beta > 0):
+        u = scalar(u, 1 / beta, b_domain=True)
+        v = AT(u)
+        alpha = normf(v)
+    else:
+        v = torch.zeros_like(x)
+        alpha = torch.zeros(1, device=device)
+
+    if torch.all(alpha > 0):
+        v = scalar(v, 1 / alpha, b_domain=False)  # v / view(alpha, Atb_shape)
```
Comment (matthieutrs) [inline deepinv/optim/utils.py:482]:
```suggestion
        v = scalar(v, 1 / alpha, b_domain=False)
```

File: deepinv/optim/utils.py [hunk 12]
```diff
@@ -49,47 +190,378 @@ def conjugate_gradient(
...
+    # n = xt.numel()/xt.size(0)
+    # var = torch.zeros_like(xt) if calc_var else None
+
+    if eta > 0:
+        if isinstance(eta, torch.Tensor):
+            eta_sqrt = torch.sqrt(eta)
+        else:
+            eta_sqrt = torch.tensor(eta, device=device).sqrt()
+
+    # ctol = 1 / conlim if conlim > 0 else 0
```
Comment (matthieutrs) [inline deepinv/optim/utils.py:446]:
```suggestion
```

File: deepinv/physics/blur.py [hunk 13]
```diff
@@ -602,7 +606,7 @@ def sinc_filter(factor=2, length=11, windowed=True):
     :param float factor: Downsampling factor.
     :param int length: Length of the filter.
     """
-    deltaf = 1 / factor
+    deltaf = 2 * (2 - 1.41) / factor
```
Comment (matthieutrs) [inline deepinv/physics/blur.py:609]: This seems a bit brutal approx for sqrt(2) to me, I would maybe go for `math.sqrt(2)` or `1.4142136`

File: deepinv/physics/forward.py [hunk 14]
```diff
@@ -479,7 +508,35 @@ def adjointness_test(self, u, **kwargs):
 
         return s1.conj() - s2
 
-    def prox_l2(self, z, y, gamma, **kwargs):
+    def condition_number(self, x, max_iter=500, tol=1e-6, verbose=False, **kwargs):
+        r"""
+        Computes approximation of the condition number of the linear operator :math:`A`.
```
Comment (matthieutrs) [inline deepinv/physics/forward.py:513]:
```suggestion
        Computes an approximation of the condition number of the linear operator :math:`A`.
```

File: deepinv/physics/remote_sensing.py [hunk 15]
```diff
@@ -90,6 +90,7 @@ def __init__(
         # Set convenience attributes
         self.downsampling = downsampling
         self.decolorize = decolorize
+        self.solver = "lsqr"  # works better than CG
```
Comment (matthieutrs) [inline deepinv/physics/remote_sensing.py:93]:
```suggestion
        self.solver = "lsqr"  # more stable than CG
```

File: deepinv/tests/test_optim.py [hunk 16]
```diff
@@ -837,3 +838,65 @@ def test_datafid_stacking(imsize, device):
...
+        < tol
+    ).all()
+
+    z = x.clone()
+    gamma = 1.0
+
+    x_hat = physics.prox_l2(z, y, gamma=gamma, solver=solver, tol=tol)
+
+    assert (
+        (x_hat - x).abs().pow(2).mean(dim=(1, 2, 3), keepdim=True)
```
Comment (matthieutrs) [inline deepinv/tests/test_optim.py:871]: if I understand well this test, it just checks that prox(x) = x, but is this the only thing we want to check?

**Review by matthieutrs — APPROVED** [review]
Thanks for the great PR, I only have few suggestions

File: deepinv/physics/blur.py (hunk 13 above)
Comment (tachella) [reply deepinv/physics/blur.py:609]: the engineer inside me wrote that

File: deepinv/tests/test_optim.py (hunk 16 above)
Comment (tachella) [reply deepinv/tests/test_optim.py:871]: I'm also checking above that the solver is working fine, ie

that |A\hat{x} - y|^2/|y|^2 is below a certain tolerance

File: deepinv/optim/utils.py [hunk 17]
```diff
@@ -29,12 +29,153 @@ def check_conv(X_prev, X, it, crit_conv="residual", thres_conv=1e-3, verbose=Fal
...
+    return dot
+
+
 def conjugate_gradient(
     A: Callable,
     b: torch.Tensor,
     max_iter: float = 1e2,
     tol: float = 1e-5,
     eps: float = 1e-8,
+    parallel_dim=None,
```
[the remark this answers is not in the document]
Comment (tachella) [reply deepinv/optim/utils.py:177]: ok
